apps/emqx_durable_storage/src/emqx_ds_storage_bitfield_lts.erl

```diff
@@ -0,0 +1,382 @@
+        )
+    ],
+    %% Obtain a keymapper for the current number of varying
+    %% levels. Magic constant 2: we have two extra dimensions of topic
+    %% index and time; the rest of dimensions are varying levels.
+    NVarying = length(Inequations) - 2,
+    Keymapper = array:get(NVarying, Keymappers),
+    Filter =
+        #filter{range_min = LowerBound, range_max = UpperBound} = emqx_ds_bitmask_keymapper:make_filter(
+            Keymapper, Inequations
+        ),
+    {ok, ITHandle} = rocksdb:iterator(DB, CF, [
+        {iterate_lower_bound, emqx_ds_bitmask_keymapper:key_to_bitstring(Keymapper, LowerBound)},
+        {iterate_upper_bound, emqx_ds_bitmask_keymapper:key_to_bitstring(Keymapper, UpperBound)}
+    ]),
+    try
+        put(?COUNTER, 0),
+        next_loop(ITHandle, Keymapper, Filter, It0, [], BatchSize)
+    after
+        rocksdb:iterator_close(ITHandle),
+        erase(?COUNTER)
+    end.
+
+%%================================================================================
+%% Internal functions
+%%================================================================================
+
+next_loop(_ITHandle, _KeyMapper, _Filter, It, Acc, 0) ->
+    {ok, It, lists:reverse(Acc)};
+next_loop(ITHandle, KeyMapper, Filter, It0, Acc0, N0) ->
+    inc_counter(),
+    #it{last_seen_key = Key0} = It0,
+    case emqx_ds_bitmask_keymapper:bin_increment(Filter, Key0) of
+        overflow ->
+            {ok, It0, lists:reverse(Acc0)};
+        Key1 ->
+            %% assert
+            true = Key1 > Key0,
+            case rocksdb:iterator_move(ITHandle, {seek, Key1}) of
+                {ok, Key, Val} ->
+                    It1 = It0#it{last_seen_key = Key},
+                    case check_message(Filter, It1, Val) of
+                        {true, Msg} ->
+                            N1 = N0 - 1,
+                            Acc1 = [Msg | Acc0];
+                        false ->
+                            N1 = N0,
+                            Acc1 = Acc0
+                    end,
+                    {N, It, Acc} = traverse_interval(ITHandle, KeyMapper, Filter, It1, Acc1, N1),
+                    next_loop(ITHandle, KeyMapper, Filter, It, Acc, N);
+                {error, invalid_iterator} ->
+                    {ok, It0, lists:reverse(Acc0)}
+            end
+    end.
+
+traverse_interval(_ITHandle, _KeyMapper, _Filter, It, Acc, 0) ->
+    {0, It, Acc};
+traverse_interval(ITHandle, KeyMapper, Filter, It0, Acc, N) ->
+    inc_counter(),
+    case rocksdb:iterator_move(ITHandle, next) of
+        {ok, Key, Val} ->
+            It = It0#it{last_seen_key = Key},
+            case check_message(Filter, It, Val) of
+                {true, Msg} ->
+                    traverse_interval(ITHandle, KeyMapper, Filter, It, [Msg | Acc], N - 1);
+                false ->
+                    traverse_interval(ITHandle, KeyMapper, Filter, It, Acc, N)
+            end;
+        {error, invalid_iterator} ->
+            {0, It0, Acc}
```

Comment:
Q: should this return `N` so that `bin_increment` in `next_loop` is tried again?  Otherwise, it seems like this `traverse_interval` fn always returns `N = 0` to `next_loop`, and then there's no need to recurse `next_loop` afterwards since it'll always match its first clause.

Reply:
> and then there's no need to recurse next_loop afterwards since it'll always match its first clause

Yes, that's the idea. When rocksDB returns `invalid_iterator` it means there are no more messages for the filter. The client must poll for the new messages later.

Reply:
It was probably not the best diff line to leave this comment, but I think the original question still stands. There currently seems to be no way for `next_loop/7` to actually loop. Anyway, I'll try to tackle that as part of EMQX-11221.

Reply:
Good catch. @keynslug is already on it.